packages/kbn-migrations/README.md

```diff
@@ -0,0 +1,322 @@
+
+- One of the seeds / transforms throws an exception
+- The index migrationState document has been tampered with and is invalid
+- If any plugins in the index's migrationState are newer than the current, active plugins (e.g. if you try to run Kibana 6.4 against a Kibana 7.0 index)
+- Connectivity errors between Kibana and Elasticsearch
+
+
+## Adding migrations to a plugin
+
+There is a yarn command to help plugin authors create new migrations:
+
+```
+yarn migration {new-seed | new-transform} {path_to_plugin_root_directory} {name_of_migration_file}
+```
+
+### Migration example
+
+As an example, let's add some migrations to the core Kibana plugin.
+
+Run `yarn migration new-seed ./src/core_plugins/kibana fancipants`, which will produce a file: `./src/core_plugins/kibana/migrations/{timestamp}_fancipants.js` as well as `./src/core_plugins/kibana/migrations/index.js`.
+
+Update the seed file to look something like this:
+
+```js
+module.exports = {
+  id: '20180430161212_seed_fancipants',
+  seed() {
+    return {
+      id: 'fanci',
+      type: 'pants',
+      attributes: {
+        description: 'Pinky in the air.',
+      },
+    };
+  }
+};
+```
+
+We also need to be sure to add `pants` to the plugin's mappings. In the case of the Kibana plugin, we need to update `./src/core_plugins/kibana/mappings.json`:
+
+```json
+{
+  "pants": {
+    "properties": {
+      "description": {
+        "type": "text"
+      }
+    }
+  },
+  // ... existing mappings omitted for brevity ...
+}
+```
+
+Now, when we start Kibana, migrations will run, and we should see a new document with id `pants:fanci` created in the `.kibana` index (or whatever the configured Kibana index is).
+
+Later in life, we may decide we want to add a new field to our `pants` documents, and we might want to ensure that all existing `pants` documents had a valid default for that field.
+
+To do that, we'd create a transform:
+
+`yarn migration new-transform ./src/core_plugins/kibana/ add_title_to_pants`
+
+That would give us a file like: `./src/core_plugins/kibana/20180430161610_transform_add_title_to_pants.js`
+
+```js
+module.exports = {
+  id: '20180430161610_transform_add_title_to_pants',
+
+  // Only apply this migration to documents of type 'pants'
+  filter: ({ type }) => 'pants',
+
+  // We're going to add a title field to these docs, so we'll
+  // give the existing docs a default value of 'N/A' for this field...
+  transform(doc) {
+    return {
+      ...doc,
+      attributes: {
+        ...doc.attributes,
+        title: 'Super Fanci',
+      },
+    };
+  },
+};
+```
+
+And again, we need a new mapping:
+
+```json
+{
+  "pants": {
+    "properties": {
+      "description": {
+        "type": "text"
+      },
+      "title": {
+        "type": "text"
+      },
+    }
+  }
+}
+```
+
+When we restart Kibana, migrations will run, and we should see that any existing `pants` documents have the title we specified in our transform.
+
+
+
+## Recovering from a failed migration
```

Comment:
While this document is mostly for developers, there are sections like this which should be on the public docs. I can work with you on finding a good home for communicating migrations and handling of failures as you have outlined here.

Here is a good starting point: https://www.elastic.co/guide/en/kibana/current/upgrade.html source is here: https://github.com/elastic/kibana/blob/master/docs/setup/upgrade.asciidoc